### ISSUE TITLE
Can't access resource variable using LookupResource in a custom op, probably because of binary incompatibility

### ISSUE THREAD

**Issue by vinowan** [tensorflow/tensorflow#39524]
**System information**

OS Platform and Distribution (e.g., Linux Ubuntu 16.04): OSX Catalina
TensorFlow installed from (source or binary): binary
TensorFlow version: 1.14.0
Python version: 3.7
Installed using virtualenv? pip? conda?: conda
GCC/Compiler version (for custom op): 

Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 11.0.3 (clang-1103.0.32.59)
Target: x86_64-apple-darwin19.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

**Describe the problem**
I have a runtime error when i run zmq_ops which is a modified version of tensorpacks's zmq_ops，and source code is in the attach files
[zmq_ops.tar.gz](https://github.com/tensorflow/tensorflow/files/4625555/zmq_ops.tar.gz)

when i run `benchmark.py` script，it raise the following errors:
2020-05-14 09:16:55.871310: W tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at zmq_pull_op.cc:61 : Invalid argument: Trying to access resource using the wrong type. Expected N10tensorpack13ZMQConnectionE got N10tensorpack13ZMQConnectionE
2020-05-14 09:16:55.871310: W tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at zmq_pull_op.cc:61 : Invalid argument: Trying to access resource using the wrong type. Expected N10tensorpack13ZMQConnectionE got N10tensorpack13ZMQConnectionE
2020-05-14 09:16:55.871357: W tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at zmq_pull_op.cc:61 : Invalid argument: Trying to access resource using the wrong type. Expected N10tensorpack13ZMQConnectionE got N10tensorpack13ZMQConnectionE
2020-05-14 09:16:55.871373: W tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at zmq_pull_op.cc:61 : Invalid argument: Trying to access resource using the wrong type. Expected N10tensorpack13ZMQConnectionE got N10tensorpack13ZMQConnectionE
2020-05-14 09:16:55.871321: W tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at zmq_pull_op.cc:61 : Invalid argument: Trying to access resource using the wrong type. Expected N10tensorpack13ZMQConnectionE got N10tensorpack13ZMQConnectionE
2020-05-14 09:16:55.871403: W tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at zmq_pull_op.cc:61 : Invalid argument: Trying to access resource using the wrong type. Expected N10tensorpack13ZMQConnectionE got N10tensorpack13ZMQConnectionE
2020-05-14 09:16:55.871421: W tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at zmq_pull_op.cc:61 : Invalid argument: Trying to access resource using the wrong type. Expected N10tensorpack13ZMQConnectionE got N10tensorpack13ZMQConnectionE
2020-05-14 09:16:55.871377: W tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at zmq_pull_op.cc:61 : Invalid argument: Trying to access resource using the wrong type. Expected N10tensorpack13ZMQConnectionE got N10tensorpack13ZMQConnectionE

issue #25113 have the same problem，but I can't find any solutions

**Comment by Saduf2019** [comment]
@vinowan 
Can you please share simple stand alone code so we could replicate the issue faced.

**Comment by vinowan** [comment]
the source code is in the attach file zmq_ops.tar.gz, you can compile and run benchmark.py scripts in it

**Comment by Saduf2019** [comment]
@vinowan 
Is it possible for you to create a  colab gist for the same and share so we could analyse the error faced.

**Comment by vinowan** [comment]
this error only occurs only in mac OSX, colab gist may not reproduce this error

**Comment by liyinhgqw** [comment]
I've got this issue too. Is it relevant to RTTI?

**Comment by liyinhgqw** [comment]
https://github.com/tensorflow/tensorflow/issues/25113

**Comment by liyinhgqw** [comment]
We've implemented a custom op using:
```
template<typename ResourceType>
Status makeResourceHandle(OpKernelContext* context, int out_idx, ResourceType **resource_) {
  static std::atomic<int64> id;
  Tensor* handle_tensor;
  TF_CHECK_STATUS(context->allocate_output(out_idx, TensorShape({}), &handle_tensor));

  ResourceType *resource = new ResourceType();
  const auto resource_name = typeid(ResourceType).name() + std::to_string(id++);
  ResourceHandle handle = MakePerStepResourceHandle<ResourceType>(context, resource_name);
  TF_CHECK_STATUS(CreateResource(context, handle, resource));
  handle_tensor->scalar<ResourceHandle>()() = handle;

  *resource_ = resource;
  return Status::OK();
}
```

Is there any chance that compiling tensorflow and our custom op on two machines would cause this issue?

**Comment by goldiegadde** [comment]
@vinowan @liyinhgqw a fix has been submitted at [1823f87](https://github.com/tensorflow/tensorflow/commit/1823f877359bb138c57a005c30aba8832dfa79fb#diff-991a6b786e16708ba1e6f5c9926cf151) and should be available in [2.3.0-rc0](https://github.com/tensorflow/tensorflow/releases/tag/v2.3.0-rc0)

**Comment by jvishnuvardhan** [comment]
@vinowan Is this still an issue? Can you please verify once and close the issue if this was already resolved. Thanks!
